### ISSUE TITLE
Check for illegal usage of functions annotated with `@TestOnly` or `@VisibleForTesting`

### ISSUE THREAD

**Issue by mrclrchtr** [detekt/detekt#5643]
## Expected Behavior of the rule
Functions annotated with [@TestOnly](https://javadoc.io/doc/org.jetbrains/annotations/latest/org/jetbrains/annotations/TestOnly.html) (JetBrains) or [@VisibleForTesting](https://guava.dev/releases/snapshot/api/docs/com/google/common/annotations/VisibleForTesting.html) (Guava) should not be used in production code, but only in test code.

## Context
I think it's the same as: https://rules.sonarsource.com/java/RSPEC-5803?search=junit

**Comment by schalkms** [comment]
This is already possible with the help of the `ForbiddenAnnotation` rule.
You could configure it like this, for instance, no need for a separate rule IMHO.

```yml
ForbiddenAnnotation:
  active: true
  excludes: ['**/test/**', '**/*.Test.kt']
  annotations:
    - value: 'org.jetbrains.annotations.TestOnly'
       reason: 'should not be used in production code, but only in test code'
    - value: 'com.google.common.annotations.VisibleForTesting'
       reason: 'should not be used in production code, but only in test code'
```

**Comment by mrclrchtr** [comment]
Hello, thank you very much for the answer. I did not know the check and I think it is great for other use cases. 

But here it is not about that the annotation should not be used, but that the function which is annotated should not be used in the production code (except in the same class, like private), but may be used in the test code.

**Comment by schalkms** [comment]
I think you are mixing up the usage of the annotation and the actual access of an annotated property or method.
- Based on your initial post, you want to prevent **the usage** of an annotation.
- Based on your recent comment, you want to prevent **the access** of an annotated property or method.
Those are two different use cases, whereas the first one is easy to achieve with the given ruleset, but the second one requires a new rule. In my opinion, it's an variation of the `ForbiddenMethodCall` rule. However, you need to fetch all elements annotated with the mentioned test annotations beforehand.

**Comment by cortinico** [comment]
> it's an variation of the `ForbiddenMethodCall` rule. However, you need to fetch all elements annotated with the mentioned test annotations beforehand.

+1 for this. On top of it, this custom rule is a bit more subtle as you'll have to access the scope from where a function is accessed. For instance, a function call within a `private` scope (same class) should be allowed, while a function call from a `public` scope should not.

**Comment by mrclrchtr** [comment]
In my initial post I did write "**functions annotated** with @TestOnly (JetBrains) or @VisibleForTesting (Guava) should not be used". But yes, I mean "to prevent access". 

Currently it is turned on in IntelliJ as an inspection error in our case, if that helps anyone reading the thread:

```
<inspection_tool class="TestOnlyProblems" enabled="true" level="INFORMATION" enabled_by_default="true" editorAttributes="INFORMATION_ATTRIBUTES">
      <scope name="Production" level="ERROR" enabled="true" editorAttributes="ERRORS_ATTRIBUTES" />
</inspection_tool>
```

However, we can't enforce it through that. We would be very happy to do it via detekt.

**Comment by BraisGabin** [comment]
🤔 This rule shouldn't be that difficult to implement. Check every function call as `ForbiddenMethodCall` does and check if that method has the forbidden annotations. Right? Ignore if the function is declared in the same file (I'm not sure how hard can this be) and then configure the rule to exclude the testing folders as other rules do.

But I'm not sure if a rule like this one is a good fit for our standard rule collection. As you said these annotations aren't coupled with kotlin itself. They are libraries. And in general we don't add anything on our standard library related with libraries. I would advice you to implement a custom rule and then link it in our [marketplace](https://detekt.dev/marketplace) because a lot of people can benefit from it.

**Comment by mrclrchtr** [comment]
>  And in general we don't add anything on our standard library related with libraries. I would advice you to implement a custom rule and then link it in our [marketplace](https://detekt.dev/marketplace) because a lot of people can benefit from it.

Ok, that is understandable. I'll see if I can find the time to implement it.

So should we close the issue then or keep it open in case it gets implemented?

**Comment by BraisGabin** [comment]
I think we can close it.
